File: packages/shared/src/types/api/career-simulation.ts
/**
 * PathForge — API Types: Career Simulation Engine™
 * ==================================================
 * Types for what-if career scenarios, projections, and comparisons.
 * Mirrors: apps/api/app/schemas/career_simulation.py
 */

// ── Simulation Sub-Components ──────────────────────────────────

export interface SimulationInputResponse {
  readonly id: string;
  readonly parameter_name: string;
  readonly parameter_value: string;
  readonly parameter_type: string;
}

export interface SimulationOutcomeResponse {
  readonly id: string;
  readonly dimension: string;
  readonly current_value: number;
  readonly projected_value: number;
  readonly delta: number;
  readonly unit: string | null;
  readonly reasoning: string | null;
}

export interface SimulationRecommendationResponse {
  readonly id: string;
  readonly priority: string;
  readonly title: string;
  readonly description: string | null;
  readonly estimated_weeks: number | null;
  readonly order_index: number;
}

// ── Full Simulation ────────────────────────────────────────────

export interface CareerSimulationResponse {
  readonly id: string;
  readonly career_dna_id: string;
  readonly scenario_type: string;
  readonly status: string;
  readonly confidence_score: number;
  readonly feasibility_rating: number;
  readonly roi_score: number | null;
  readonly salary_impact_percent: number | null;
  readonly estimated_months: number | null;
  readonly reasoning: string | null;
  readonly factors: Record<string, unknown> | null;
  readonly data_source: string;
  readonly disclaimer: string;
  readonly computed_at: string;
  readonly inputs: SimulationInputResponse[];
  readonly outcomes: SimulationOutcomeResponse[];
  readonly recommendations: SimulationRecommendationResponse[];
}

// ── Summary (for lists/dashboard) ──────────────────────────────

export interface SimulationSummaryResponse {
  readonly id: string;
  readonly scenario_type: string;
  readonly status: string;
  readonly confidence_score: number;
  readonly salary_impact_percent: number | null;
  readonly estimated_months: number | null;
  readonly data_source: string;
  readonly disclaimer: string;
  readonly computed_at: string;
}

// ── Preferences ────────────────────────────────────────────────

export interface SimulationPreferenceResponse {
  readonly id: string;
  readonly career_dna_id: string;
  readonly default_scenario_type: string | null;
  readonly max_scenarios: number;
  readonly notification_enabled: boolean;
}

export interface SimulationPreferenceUpdateRequest {
  readonly default_scenario_type?: string | null;
  readonly max_scenarios?: number;
  readonly notification_enabled?: boolean;
}

// ── Comparison ─────────────────────────────────────────────────

export interface SimulationComparisonResponse {
  readonly simulations: CareerSimulationResponse[];
  readonly ranking: string[];
  readonly trade_off_analysis: string | null;
  readonly data_source: string;
  readonly disclaimer: string;
}

// ── Dashboard ──────────────────────────────────────────────────

export interface SimulationDashboardResponse {
  readonly simulations: SimulationSummaryResponse[];
  readonly preferences: SimulationPreferenceResponse | null;
  readonly total_simulations: number;
  readonly scenario_type_counts: Record<string, number>;
}

// ── Request Schemas ────────────────────────────────────────────

export interface RoleTransitionSimRequest {
  readonly target_role: string;
  readonly target_industry?: string | null;
  readonly target_location?: string | null;
}

export interface GeoMoveSimRequest {
  readonly target_location: string;
  readonly keep_role?: boolean;
  readonly target_role?: string | null;
}

export interface SkillInvestmentSimRequest {
  readonly skills: string[];
  readonly target_role?: string | null;
}

export interface IndustryPivotSimRequest {
  readonly target_industry: string;
  readonly target_role?: string | null;
}

export interface SeniorityJumpSimRequest {
  readonly target_seniority: string;
  readonly target_role?: string | null;
}

export interface SimulationCompareRequest {
  readonly simulation_ids: string[];
}

File: apps/api/app/schemas/career_simulation.py
"""
PathForge — Career Simulation Engine™ Schemas
================================================
Request/response Pydantic schemas for the Career Simulation API.

Request schemas validate user input with constraints.
Response schemas provide full transparency (data_source, disclaimer).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# ── Response Schemas ─────────────────────────────────────────


class SimulationInputResponse(BaseModel):
    """A single input parameter key-value pair."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    parameter_name: str
    parameter_value: str
    parameter_type: str


class SimulationOutcomeResponse(BaseModel):
    """A single dimension projection (current → projected → delta)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    dimension: str
    current_value: float
    projected_value: float
    delta: float
    unit: str | None = None
    reasoning: str | None = None


class SimulationRecommendationResponse(BaseModel):
    """A single actionable recommendation with priority."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    priority: str
    title: str
    description: str | None = None
    estimated_weeks: int | None = None
    order_index: int


class CareerSimulationResponse(BaseModel):
    """Full simulation detail with all projections and recommendations."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    career_dna_id: uuid.UUID
    scenario_type: str
    status: str
    confidence_score: float
    feasibility_rating: float
    roi_score: float | None = None
    salary_impact_percent: float | None = None
    estimated_months: int | None = None
    reasoning: str | None = None
    factors: dict[str, Any] | None = None
    data_source: str
    disclaimer: str
    computed_at: datetime
    inputs: list[SimulationInputResponse] = Field(default_factory=list)
    outcomes: list[SimulationOutcomeResponse] = Field(default_factory=list)
    recommendations: list[SimulationRecommendationResponse] = Field(default_factory=list)


class SimulationSummaryResponse(BaseModel):
    """Lightweight simulation card for list views and dashboard."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    scenario_type: str
    status: str
    confidence_score: float
    salary_impact_percent: float | None = None
    estimated_months: int | None = None
    data_source: str
    disclaimer: str
    computed_at: datetime


class SimulationPreferenceResponse(BaseModel):
    """User preferences for the simulation module."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    career_dna_id: uuid.UUID
    default_scenario_type: str | None = None
    max_scenarios: int
    notification_enabled: bool


class SimulationComparisonResponse(BaseModel):
    """Side-by-side comparison of multiple simulations."""

    model_config = ConfigDict(from_attributes=True)

    simulations: list[CareerSimulationResponse]
    ranking: list[uuid.UUID] = Field(
        default_factory=list,
        description="Simulation IDs ranked by composite desirability",
    )
    trade_off_analysis: str | None = None
    data_source: str = "AI-generated comparison based on Career DNA and market data"
    disclaimer: str = (
        "This comparison is an AI-generated analysis, not a guarantee. "
        "Actual outcomes depend on market conditions, personal effort, "
        "and factors beyond prediction."
    )


class SimulationDashboardResponse(BaseModel):
    """Dashboard: all simulations + preferences + summary stats."""

    simulations: list[SimulationSummaryResponse] = Field(default_factory=list)
    preferences: SimulationPreferenceResponse | None = None
    total_simulations: int = 0
    scenario_type_counts: dict[str, int] = Field(default_factory=dict)


# ── Request Schemas ──────────────────────────────────────────


class RoleTransitionSimRequest(BaseModel):
    """What-if: role transition scenario."""

    target_role: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Target role to simulate transition to",
    )
    target_industry: str | None = Field(
        None,
        max_length=255,
        description="Optional target industry for context",
    )
    target_location: str | None = Field(
        None,
        max_length=255,
        description="Optional target location for context",
    )


class GeoMoveSimRequest(BaseModel):
    """What-if: geographic relocation scenario."""

    target_location: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Target location to simulate moving to",
    )
    keep_role: bool = Field(
        True,
        description="Whether to keep the same role after the move",
    )
    target_role: str | None = Field(
        None,
        max_length=255,
        description="Optional new role in the target location",
    )


class SkillInvestmentSimRequest(BaseModel):
    """What-if: skill investment scenario."""

    skills: list[str] = Field(
        ...,
        min_length=1,
        max_length=10,
        description="Skills to simulate investing in (1-10)",
    )
    target_role: str | None = Field(
        None,
        max_length=255,
        description="Optional target role that benefits from these skills",
    )


class IndustryPivotSimRequest(BaseModel):
    """What-if: industry pivot scenario."""

    target_industry: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Target industry to pivot to",
    )
    target_role: str | None = Field(
        None,
        max_length=255,
        description="Optional target role in that industry",
    )


class SeniorityJumpSimRequest(BaseModel):
    """What-if: seniority level change scenario."""

    target_seniority: str = Field(
        ...,
        min_length=2,
        max_length=100,
        description="Target seniority level (e.g. 'Lead', 'Staff', 'Director')",
    )
    target_role: str | None = Field(
        None,
        max_length=255,
        description="Optional specific role at that seniority",
    )


class SimulationCompareRequest(BaseModel):
    """Request to compare multiple saved simulations."""

    simulation_ids: list[uuid.UUID] = Field(
        ...,
        min_length=2,
        max_length=5,
        description="2-5 simulation IDs to compare side-by-side",
    )


class SimulationPreferenceUpdateRequest(BaseModel):
    """Partial update for simulation preferences."""

    default_scenario_type: str | None = None
    max_scenarios: int | None = Field(None, ge=1, le=100)
    notification_enabled: bool | None = None
